### DISCUSSION TITLE
BlockingQueue는 어떤 자료구조일까?

### DISCUSSION THREAD

**Discussion by kmularise** [dailyge/dailyge-server#121] [Information]
## 📝 연관 PR

<br/><br/>

## ❓ 작성 이유
사용자의 요청이 일어날때마다, Redis나 DB를 호출하게 되면 네트워크 호출이 일어나 응답속도가 떨어지게 됩니다. 따라서 사용자의 요청이 일어나면 내부의 메모리에 저장해두었다가, 요청 벌크가 쌓이면 그 때 Redis나 DB를 호출해서 한꺼번에 처리하는 것이 효율적입니다. 이때, 주의할 점이 먼저 일어난 요청이 먼저 처리되어야 하므로 큐를 사용해야 하고, 스레드간 동기화가 지원되는 큐 자료구조를 사용해야 한다는 것입니다. 이러한 자료구조는 Java에서 BlockingQueue라는 인터페이스의 구현체들로 구현되어 있습니다. BlockingQueue와 대표적인 구현체인 LinkedBlockingQueue, ArrayBlcokingQueue에 대해서 알아보기 위해 이글을 작성했습니다.

<br/><br/>

## 1. BlockingQueue와 BlockingQueue의 구현체들
BlockingQueue는 멀티스레드 환경에서 안전하게 사용될 수 있도록 설계된 큐 자료구조 입니다. BlockingQueue는 스레드 동기화를 지원하기 위해 다음과 같은 방법들을 사용합니다.
<br>
- `ReentrantLock`과 같은 락을 사용합니다.
- 내부에서 size와 같은 변수들을 `AtomicInteger` 같은 원자성을 보장하는 클래스를 사용합니다.

다음으로는 대표적인 BlockingQueue 구현체인 LinkedBlockingQueue와 ArrayBlcokingQueue를 알아보겠습니다.

### 1-1. LinkedBlockingQueue
- 연결된 노드를 기반으로 구현됩니다.
- 용량(capacity)에 제한이 없습니다.
- 노드를 미리 할당하지 않으므로 할당된 메모리 크기와 메모리 사용량이 대체로 유사합니다.
- 각 노드가 메모리 임의의 위치에 있을 수 있기에 캐시 히트율이 낮습니다.
- 요소 삽입과 삭제 작업에서 별도의 락을 사용합니다. 
<br/><br/>

### 1-2. ArrayBlcokingQueue
- 배열을 기반으로 구현됩니다.
- 용량(capacity)에 제한이 있습니다.
- 배열을 미리 할당하므로 상당한 메모리를 차지할 수 있습니다.
- 메모리 접근이 연속적으로 일어나 캐시 히트율이 높습니다.
- 요소 삽입과 삭제 작업이 단일 락으로 관리됩니다.
<br/><br/>

## 2. 프로젝트에서 선택한 자료구조
이번 프로젝트에서는 인메모리에 대한 자료구조로 LinkedBlockingQueue을 선택하기로 했습니다. 
<br>
그 이유는 
- 용량(capacity)에 제한이 없기에 용량 부족으로 예외를 발생시키거나 스레드가 대기하는 일이 ArrayBlockingQueue보다 없기 때문입니다.
- 요소 삽입 때 메모리를 할당할 수 있기 때문에 메모리 사용량을 유동적으로 관리할 수 있기 때문입니다.

추후에 실제 테스트를 통해 LinkedBlockingQueue와 ArrayBlockingQueue를 비교해볼 수 있을 것 같습니다.

**Comment by devjun10** [comment]
좋네요. 성능 면에서는 어떤 자료구조가 더 우수한가요?

**Reply by kmularise** [reply]
큐 자료구조에 한해서는 LinkedBlockingQueue가 더 우수하다고 생각합니다.
<br>
그 이유는 다음과 같습니다.
1. LinkedBlockingQueue에서는 용량 제한이 없기 때문에, 큐가 다 찼을 때 스레드가 대기하는 경우를 없게끔 설정할 수 있습니다.
2. LinkedBlockingQueue는 큐의 앞쪽과 뒤쪽에 별도의 락을 걸기 때문에 동시에 앞쪽에서 요소를 제거하면서 맨뒤쪽으로 요소를 넣을 수 있기 때문에 단일 락으로 제어하는 ArrayBlockingQueue보다 병렬적으로 작업을 처리할 수 있습니다.
